### ISSUE TITLE
New Source: Twitter Ads

### ISSUE THREAD

**Issue by dubravcik** [airbytehq/airbyte#3107]
## Tell us about the new integration you’d like to have
There are few mainstream marketing platforms (Google Ads, LinkedIn, Facebook, Bing etc) and Twitter is one of them. We need to analyze performance of ads like impressions, conversions, costs etc per campaign/ad scheduled in our company account.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368227853339) by [Unito](https://www.unito.io)

**Comment by sherifnada** [comment]
@dubravcik Yup, this is definitely a great one to have. Would you be able to point us to which endpoints in the API you'd like to support off the bat? https://developer.twitter.com/en/docs/twitter-ads-api/api-reference-index

There's quite a few endpoints so it's easy to get lost 😅

**Comment by yaroslav-dudar** [comment]
Integration Vetting
---------------------------

**Webhook-based? (no/partially/yes)**
No

**Available authentication modes (API key/Oauth/other)**
API keys - for own Twitter Ads Account (single account use case)
OAuth - to use another Twitter Ads Account

```python
CONSUMER_KEY = 'your consumer key'
CONSUMER_SECRET = 'your consumer secret'
ACCESS_TOKEN = 'access token'
ACCESS_TOKEN_SECRET = 'access token secret'
ACCOUNT_ID = 'account id'
```
By default, each Twitter developer application contains a static access token that can be used to make Ads API requests for the account that owns the application. These credentials are ideal for single-account use cases without requiring a 3-legged or PIN-based OAuth flow. If you aren't accessing another Twitter Ads Account, use these single-user credentials

**Creating an account**
Require these steps:
- Twitter Developer account (require approval) https://developer.twitter.com/en/apply-for-access
- Apply for access to the Ads API (require approval) https://developer.twitter.com/en/docs/twitter-ads-api/apply

**Has Sandbox environment ?**
Yes. https://developer.twitter.com/en/docs/twitter-ads-api/ads-api-sandbox

**How to populate the account with data?**
It can be done through the UI, API, Sandbox API.

**Available streams for sync**
https://developer.twitter.com/en/docs/twitter-ads-api/api-reference-index

**Integration supports incremental sync?**
Yes, API has pagination support , so it's possible to save cursor id for future use
https://developer.twitter.com/en/docs/twitter-ads-api/pagination
https://developer.twitter.com/en/docs/pagination

**Other information/blockers**
Require approval from Twitter to create account.

To perform actions on behalf of another user, you'll need to obtain their access tokens via 3-legged OAuth flow.
https://developer.twitter.com/en/docs/authentication/oauth-1-0a/obtaining-user-access-tokens

Twitter ads account (not sandbox) require payment: https://business.twitter.com/en/help/account-setup/ads-account-creation.html
Billing - https://business.twitter.com/en/help/account-setup/billing-basics/credit-limits.html

`API call limits `- https://developer.twitter.com/en/docs/twitter-ads-api/rate-limiting
`Python SDK` - https://github.com/twitterdev/twitter-python-ads-sdk

**Comment by cheyura** [comment]
Hi,
Thanks for that.I will add if smith is relevant

Ср, 2 июня 2021 г. в 15:06, Yaroslav Dudar ***@***.***>:

**Comment by sherifnada** [comment]
@cheyura putting this back into Vetting In progress - can we do the same thing we did with Google Ads for this one? Starting a campaign with ~$20/mo if possible. You'll need to request access. Please use the integration-test@airbyte.io email address for any sign ups.

**Comment by sherifnada** [comment]
@cheyura what is the blocker?

**Comment by cheyura** [comment]
No blocker for now.Sorry forgot to move

пн, 7 черв. 2021, 21:58 користувач Sherif A. Nada ***@***.***>
пише:

**Comment by midavadim** [comment]
Core streams (subset of stitch data): 
    Accounts
    Account media
    Campaigns
    Tweets
    Promoted Tweets
    
Report Streams (subset of Fivetran’s reports):

- Metrics reports:

    Account report            (14 sub streams with segmentation breakdowns)
    Campaign report           (17 sub streams with segmentation breakdowns)
    Funding Instrument report (15 sub streams with segmentation breakdowns)
    Line Item report          (16 sub streams with segmentation breakdowns)
    Promoted tweet report     (16 sub streams with segmentation breakdowns)

    Each of these reports should be available as 14 different streams, each corresponding to one of the 13 available segmentation breakdowns (see Fivetran ERD), and the 14th one should correspond to having no breakdows. 

- Reach Reports
    Funding Instrument report (FUNDING_INSTRUMENT_REACH_DAILY_REPORT, FUNDING_INSTRUMENT_REACH_MONTHLY_REPORT)
    Campaign report                     (CAMPAIGN_REACH_DAILY_REPORT,           CAMPAIGN_REACH_MONTHLY_REPORT)
    
    Each of these reports should be available as three streams, each corresponding to an aggregation strategy (see fivetran ERD) 

References
Stitch Data docs: 
    https://www.stitchdata.com/docs/integrations/saas/twitter-ads#custom-report-configuration-options 
Fivetran Docs: 
    https://fivetran.com/docs/applications/twitter-ads

**Comment by TymoshokDmytro** [comment]
As @yaroslav-dudar said in the vetting comment: 

Creating an account requires these steps:
- Twitter Developer account (require approval) https://developer.twitter.com/en/apply-for-access
- Apply for access to the Ads API (require approval) https://developer.twitter.com/en/docs/twitter-ads-api/apply

The first is done. The second is still on Twitter review from 13th of July.

**Comment by mike-wax** [comment]
Looks like this one is done, any idea when it'll be merged? Is there an easy way to use these ahead of GA, happy to help test.

**Comment by ltieman** [comment]
Hello,

Was wanting to know what the timeline was on this?  Also happy to help with testing etc.

**Comment by darian-heede** [comment]
Hi, I'm also interested in knowing what the state is here? Is there something blocking the merge?

**Comment by sherifnada** [comment]
Hi everyone, this connector is not currently on our roadmap as we're focused on reliability and performance of the existing set of connectors. We're happy to help support a contribution if anyone wants to make a PR! Will keep you all updated about roadmap status as time goes on.

**Comment by YowanR** [comment]
FYI that we should instead use API v11 or above instead of v10 when we tackle this one. https://twittercommunity.com/t/ads-api-version-11/168814?utm_campaign=ADS_PU_GBL_EN_V10_InitialNotice_Sunset+Notification_220721&utm_medium=email&utm_source=Eloqua&ref=

**Comment by ycherniaiev** [comment]
## Tell us about the problem you're trying to solve
**DOCS**
[Twiiter Ads API Reference](https://developer.twitter.com/en/docs/twitter-ads-api)

**Sandbox Account Creds**
https://ads.twitter.com with LastPass creds: Twitter - @AirbyteHQ - for integration tests

**OAuth 1.0 and Bearer Token:**
LastPass: source Twitter Ads Developer App

**Vetting**
- [ ] #3107

**MLP**
[Document](https://docs.google.com/document/d/1Q2wvM-nlYUAZLY_gN-EjFCpJtB8dbv4cMv9qvMs5_3s/edit)

**Comment by YowanR** [comment]
For this one, pasting @girarda's initial assessment of the connector: "seems simple except for the asynchronous analytics endpoints, which we can't write in YAML this quarter". I think that we can proceed with this one.
cc @sherifnada in case you want to chime in

**Comment by roman-yermilov-gl** [comment]
Hi @YowanR 
I made some investigation during scoping and this is some findings:

It was requested to implement this connector using YAML low-code but is not possible for now because of asynchronous API for analytics.
Using YAML low-code we can implement only synchronous API but with some restrictions:
Functional restrictions:
    - we will have small date range window of 7 days instead of 45 days (or even 90 days in case when segmentation is not specified)
    - we will not be able to use segmentation. It means we can propose only 1 stream while fivetran proposes 14 for every available Entity (there are 7 Entities) in the system
    - we will not be able to use all the granuality types. It means we can propose only 2 streams while fivetran proposes 3 for Reach Reports (there 2 types of such a reports)
Technical question:
    - Twitter supports OAuth 1.0 but we don't have it for any other connector. Do we need it to support only one connector when Twitter says they working on OAuth 2.0 for Twitter Ads? https://twittercommunity.com/t/oauth-2-0-support-for-twitter-ads-api/172165
For this reasons I think low-code solution is not even close to what other platforms propose to their customers. So another way is to implement oldschool Python connector with some kind of scheduler or polling for asynchronous API or just postpone it until better times

**Comment by YowanR** [comment]
I wasn't aware of these additional restrictions -- thanks for looking into it! I wonder if @girarda or @sherifnada have thoughts on the restrictions such as the date range (way smaller than Python). These could be blockers for adoption, IMO.

**Comment by sherifnada** [comment]
I agree with roman's suggestion. This connector is not a good fit for lowcode

**Comment by roman-yermilov-gl** [comment]
@sherifnada @YowanR 
This is scoping for Twitter Ads alpha which covers what we can do with low-code and synchronous API: https://docs.google.com/document/d/1m9hVYyJPpE1JHYdDd73_RXiLb94nwvatEJ-DpYG-11Y

**Comment by YowanR** [comment]
I don't think we should close this. We are still planning to get Twitter Ads supported at Airbyte but it might not be through the current version of the low-code framework. Please keep this GH issues open

**Comment by brunopini** [comment]
Hey @lazebnyi, just wondering if there are any updates on this? I'll need to work with Twitter Ads data for a project next quarter. From reading this issue, I guess I shouldn't even bother trying the no-code implementation via Airbyte, and rather develop my own external Python connector until there is a native solution from you guys?
